Test failure on Python 3.13 because it strips indents from docstrings

## Expected behavior and actual behavior.

In Python 3.13 (currently in alpha), as described in https://docs.python.org/3.13/whatsnew/3.13.html, “Compiler now strip indents from docstrings.” See https://github.com/python/cpython/issues/81283.

There doesn’t yet seem to be a working Python 3.13 wheel for `numpy` on PyPI, which makes this awkward to reproduce in a virtualenv, but we are already doing [early Python 3.13 testing](https://copr.fedorainfracloud.org/coprs/g/python/python3.13/) in Fedora Linux.

As reported in https://bugzilla.redhat.com/show_bug.cgi?id=2250678, some tests fail due to the new docstring handling.

```
_______________________ test_requires_geos_method[3.8.0] _______________________

version = '3.8.0', mocked_geos_version = '3.7.1', sphinx_doc_build = None

    @pytest.mark.parametrize("version", ["3.6.0", "3.8.0"])
    def test_requires_geos_method(version, mocked_geos_version, sphinx_doc_build):
        """The requires_geos decorator adjusts methods docstrings correctly"""
        wrapped = requires_geos(version)(SomeClass.func)
    
>       assert wrapped.__doc__ == expected_docstring.format(version=version, indent=" " * 8)
E       AssertionError: assert 'Docstring th...escription.\n' == 'Docstring th...on.\n        '
E           Docstring that will be mocked.
E         -         A multiline.
E         ? --------
E         + A multiline.
E           
E         -         .. note:: 'func' requires at least GEOS 3.8.0.
E         ? ------...
E         
E         ...Full output truncated (6 lines hidden), use '-vv' to show

tests/test_misc.py:123: AssertionError
=========================== short test summary info ============================
FAILED tests/test_misc.py::test_requires_geos_not_ok[3.7.2] - AssertionError:...
FAILED tests/test_misc.py::test_requires_geos_not_ok[3.8.0] - AssertionError:...
FAILED tests/test_misc.py::test_requires_geos_not_ok[3.8.1] - AssertionError:...
FAILED tests/test_misc.py::test_requires_geos_doc_build[3.6.0] - AssertionErr...
FAILED tests/test_misc.py::test_requires_geos_doc_build[3.8.0] - AssertionErr...
FAILED tests/test_misc.py::test_requires_geos_method[3.6.0] - AssertionError:...
FAILED tests/test_misc.py::test_requires_geos_method[3.8.0] - AssertionError:...
============ 7 failed, 3494 passed, 37 skipped, 3 xfailed in 5.74s =============
```

We expect that all tests should pass.

## Steps to reproduce the problem.

Compile Shapely and run the tests on Python 3.13. (The hard part is getting a proper installation of Numpy.)

## Operating system

Fedora Linux Rawhide – early testing of Python 3.13.

## Shapely version and provenance

2.0.2, [`python-shapey`](https://src.fedoraproject.org/rpms/python-shapely)

-----

I intend to work on a PR for this.

Output with `pytest -vv`:

```
=================================== FAILURES ===================================
_______________________ test_requires_geos_not_ok[3.7.2] _______________________

version = '3.7.2', mocked_geos_version = '3.7.1'

    @pytest.mark.parametrize("version", ["3.7.2", "3.8.0", "3.8.1"])
    def test_requires_geos_not_ok(version, mocked_geos_version):
        wrapped = requires_geos(version)(func)
        with pytest.raises(shapely.errors.UnsupportedGEOSVersionError):
            wrapped()

>       assert wrapped.__doc__ == expected_docstring.format(version=version, indent=" " * 4)
E       assert "Docstring that will be mocked.\nA multiline.\n\n  .. note:: 'func' requires at least GEOS 3.7.2.\n\nSome description.\n" == "Docstring that will be mocked.\n    A multiline.\n\n    .. note:: 'func' requires at least GEOS 3.7.2.\n\n    Some description.\n    "
E           Docstring that will be mocked.
E         -     A multiline.
E         ? ----
E         + A multiline.
E
E         -     .. note:: 'func' requires at least GEOS 3.7.2.
E         ? --
E         +   .. note:: 'func' requires at least GEOS 3.7.2.
E
E         -     Some description.
E         ? ----
E         + Some description.
E         -

tests/test_misc.py:107: AssertionError
_______________________ test_requires_geos_not_ok[3.8.0] _______________________

version = '3.8.0', mocked_geos_version = '3.7.1' 

    @pytest.mark.parametrize("version", ["3.7.2", "3.8.0", "3.8.1"])
    def test_requires_geos_not_ok(version, mocked_geos_version):
        wrapped = requires_geos(version)(func)   
        with pytest.raises(shapely.errors.UnsupportedGEOSVersionError):
            wrapped()

>       assert wrapped.__doc__ == expected_docstring.format(version=version, indent=" " * 4)
E       assert "Docstring that will be mocked.\nA multiline.\n\n  .. note:: 'func' requires at least GEOS 3.8.0.\n\nSome description.\n" == "Docstring that will be mocked.\n    A multiline.\n\n    .. note:: 'func' requires at least GEOS 3.8.0.\n\n    Some description.\n    "
E           Docstring that will be mocked.
E         -     A multiline.
E         ? ----
E         + A multiline.
E
E         -     .. note:: 'func' requires at least GEOS 3.8.0.
E         ? --
E         +   .. note:: 'func' requires at least GEOS 3.8.0.
E
E         -     Some description.
E         ? ----
E         + Some description.
E         -

tests/test_misc.py:107: AssertionError
_______________________ test_requires_geos_not_ok[3.8.1] _______________________

version = '3.8.1', mocked_geos_version = '3.7.1' 

    @pytest.mark.parametrize("version", ["3.7.2", "3.8.0", "3.8.1"])
    def test_requires_geos_not_ok(version, mocked_geos_version):
        wrapped = requires_geos(version)(func)   
        with pytest.raises(shapely.errors.UnsupportedGEOSVersionError):
            wrapped()

>       assert wrapped.__doc__ == expected_docstring.format(version=version, indent=" " * 4)
E       assert "Docstring that will be mocked.\nA multiline.\n\n  .. note:: 'func' requires at least GEOS 3.8.1.\n\nSome description.\n" == "Docstring that will be mocked.\n    A multiline.\n\n    .. note:: 'func' requires at least GEOS 3.8.1.\n\n    Some description.\n    "
E           Docstring that will be mocked.
E         -     A multiline.
E         ? ----
E         + A multiline.
E
E         -     .. note:: 'func' requires at least GEOS 3.8.1.
E         ? --
E         +   .. note:: 'func' requires at least GEOS 3.8.1.
E
E         -     Some description.
E         ? ----
E         + Some description.
E         -

tests/test_misc.py:107: AssertionError
_____________________ test_requires_geos_doc_build[3.6.0] ______________________

version = '3.6.0', mocked_geos_version = '3.7.1', sphinx_doc_build = None

    @pytest.mark.parametrize("version", ["3.6.0", "3.8.0"])
    def test_requires_geos_doc_build(version, mocked_geos_version, sphinx_doc_build):
        """The requires_geos decorator always adapts the docstring."""
        wrapped = requires_geos(version)(func)   

>       assert wrapped.__doc__ == expected_docstring.format(version=version, indent=" " * 4)
E       assert "Docstring that will be mocked.\nA multiline.\n\n  .. note:: 'func' requires at least GEOS 3.6.0.\n\nSome description.\n" == "Docstring that will be mocked.\n    A multiline.\n\n    .. note:: 'func' requires at least GEOS 3.6.0.\n\n    Some description.\n    "
E           Docstring that will be mocked.
E         -     A multiline.
E         ? ----
E         + A multiline.
E
E         -     .. note:: 'func' requires at least GEOS 3.6.0.
E         ? --
E         +   .. note:: 'func' requires at least GEOS 3.6.0.
E
E         -     Some description.
E         ? ----
E         + Some description.
E         -

tests/test_misc.py:115: AssertionError
_____________________ test_requires_geos_doc_build[3.8.0] ______________________

version = '3.8.0', mocked_geos_version = '3.7.1', sphinx_doc_build = None

    @pytest.mark.parametrize("version", ["3.6.0", "3.8.0"])
    def test_requires_geos_doc_build(version, mocked_geos_version, sphinx_doc_build):
        """The requires_geos decorator always adapts the docstring."""
        wrapped = requires_geos(version)(func)   

>       assert wrapped.__doc__ == expected_docstring.format(version=version, indent=" " * 4)
E       assert "Docstring that will be mocked.\nA multiline.\n\n  .. note:: 'func' requires at least GEOS 3.8.0.\n\nSome description.\n" == "Docstring that will be mocked.\n    A multiline.\n\n    .. note:: 'func' requires at least GEOS 3.8.0.\n\n    Some description.\n    "
E           Docstring that will be mocked.
E         -     A multiline.
E         ? ----
E         + A multiline.
E
E         -     .. note:: 'func' requires at least GEOS 3.8.0.
E         ? --
E         +   .. note:: 'func' requires at least GEOS 3.8.0.
E
E         -     Some description.
E         ? ----
E         + Some description.
E         -

tests/test_misc.py:115: AssertionError
_______________________ test_requires_geos_method[3.6.0] _______________________

version = '3.6.0', mocked_geos_version = '3.7.1', sphinx_doc_build = None

    @pytest.mark.parametrize("version", ["3.6.0", "3.8.0"])
    def test_requires_geos_method(version, mocked_geos_version, sphinx_doc_build):
        """The requires_geos decorator adjusts methods docstrings correctly"""
        wrapped = requires_geos(version)(SomeClass.func)

>       assert wrapped.__doc__ == expected_docstring.format(version=version, indent=" " * 8)
E       assert "Docstring that will be mocked.\nA multiline.\n\n  .. note:: 'func' requires at least GEOS 3.6.0.\n\nSome description.\n" == "Docstring that will be mocked.\n        A multiline.\n\n        .. note:: 'func' requires at least GEOS 3.6.0.\n\n        Some description.\n        "
E           Docstring that will be mocked.
E         -         A multiline.
E         ? --------
E         + A multiline.
E
E         -         .. note:: 'func' requires at least GEOS 3.6.0.
E         ? ------
E         +   .. note:: 'func' requires at least GEOS 3.6.0.
E
E         -         Some description.
E         ? --------
E         + Some description.
E         -

tests/test_misc.py:123: AssertionError
_______________________ test_requires_geos_method[3.8.0] _______________________

version = '3.8.0', mocked_geos_version = '3.7.1', sphinx_doc_build = None

    @pytest.mark.parametrize("version", ["3.6.0", "3.8.0"])
    def test_requires_geos_method(version, mocked_geos_version, sphinx_doc_build):
        """The requires_geos decorator adjusts methods docstrings correctly"""
        wrapped = requires_geos(version)(SomeClass.func)

>       assert wrapped.__doc__ == expected_docstring.format(version=version, indent=" " * 8)
E       assert "Docstring that will be mocked.\nA multiline.\n\n  .. note:: 'func' requires at least GEOS 3.8.0.\n\nSome description.\n" == "Docstring that will be mocked.\n        A multiline.\n\n        .. note:: 'func' requires at least GEOS 3.8.0.\n\n        Some description.\n        "
E           Docstring that will be mocked.
E         -         A multiline.
E         ? --------
E         + A multiline.
E
E         -         .. note:: 'func' requires at least GEOS 3.8.0.
E         ? ------
E         +   .. note:: 'func' requires at least GEOS 3.8.0.
E
E         -         Some description.
E         ? --------
E         + Some description.
E         -

tests/test_misc.py:123: AssertionError
=========================== short test summary info ============================
FAILED tests/test_misc.py::test_requires_geos_not_ok[3.7.2] - assert "Docstri...
FAILED tests/test_misc.py::test_requires_geos_not_ok[3.8.0] - assert "Docstri...
FAILED tests/test_misc.py::test_requires_geos_not_ok[3.8.1] - assert "Docstri...
FAILED tests/test_misc.py::test_requires_geos_doc_build[3.6.0] - assert "Docs...
FAILED tests/test_misc.py::test_requires_geos_doc_build[3.8.0] - assert "Docs...
FAILED tests/test_misc.py::test_requires_geos_method[3.6.0] - assert "Docstri...
FAILED tests/test_misc.py::test_requires_geos_method[3.8.0] - assert "Docstri...
============ 7 failed, 3494 passed, 37 skipped, 3 xfailed in 5.12s =============
```

--- a/shapely/decorators.py
+++ b/shapely/decorators.py
@@ -39,7 +39,7 @@ class requires_geos:
             # Insert the message at the first double newline
             position = doc.find("\n\n") + 2
             # Figure out the indentation level
-            indent = 2
+            indent = 0
             while True:
                 if doc[position + indent] == " ":
                     indent += 1

--- a/shapely/tests/test_misc.py
+++ b/shapely/tests/test_misc.py
@@ -87,18 +87,18 @@ def expected_docstring(**kwds):
     doc = """Docstring that will be mocked.
 {indent}A multiline.
 
-{indent}{extra_indent}.. note:: 'func' requires at least GEOS {version}.
+{indent}.. note:: 'func' requires at least GEOS {version}.
 
 {indent}Some description.
-{indent}"""
+{indent}""".format(
+        **kwds
+    )
     if sys.version_info[:2] >= (3, 13):
         # There are subtle differences between inspect.cleandoc() and
         # _PyCompile_CleanDoc(). Most significantly, the latter does not remove
-        # leading or trailing blank lines. The extra_indent is required because
-        # the algorithm in shapely.decorators.requires_geos assumes a minimum
-        # indent of two spaces.
-        return cleandoc(doc.format(extra_indent="  ", **kwds)) + "\n"
-    return doc.format(extra_indent="", **kwds)
+        # leading or trailing blank lines.
+        return cleandoc(doc) + "\n"
+    return doc
 
 
 @pytest.mark.parametrize("version", ["3.10.0", "3.10.1", "3.9.2"])

--- a/shapely/tests/test_misc.py
+++ b/shapely/tests/test_misc.py
@@ -1,5 +1,6 @@
 import os
 import sys
+from inspect import cleandoc
 from itertools import chain
 from string import ascii_letters, digits
 from unittest import mock
@@ -82,13 +83,22 @@ class SomeClass:
         """
 
 
-expected_docstring = """Docstring that will be mocked.
+def expected_docstring(**kwds):
+    doc = """Docstring that will be mocked.
 {indent}A multiline.
 
-{indent}.. note:: 'func' requires at least GEOS {version}.
+{indent}{extra_indent}.. note:: 'func' requires at least GEOS {version}.
 
 {indent}Some description.
 {indent}"""
+    if sys.version_info[:2] >= (3, 13):
+        # There are subtle differences between inspect.cleandoc() and
+        # _PyCompile_CleanDoc(). Most significantly, the latter does not remove
+        # leading or trailing blank lines. The extra_indent is required because
+        # the algorithm in shapely.decorators.requires_geos assumes a minimum
+        # indent of two spaces.
+        return cleandoc(doc.format(extra_indent="  ", **kwds)) + "\n"
+    return doc.format(extra_indent="", **kwds)
 
 
 @pytest.mark.parametrize("version", ["3.10.0", "3.10.1", "3.9.2"])
@@ -104,7 +114,7 @@ def test_requires_geos_not_ok(version, mocked_geos_version):
     with pytest.raises(shapely.errors.UnsupportedGEOSVersionError):
         wrapped()
 
-    assert wrapped.__doc__ == expected_docstring.format(version=version, indent=" " * 4)
+    assert wrapped.__doc__ == expected_docstring(version=version, indent=" " * 4)
 
 
 @pytest.mark.parametrize("version", ["3.9.0", "3.10.0"])
@@ -112,7 +122,7 @@ def test_requires_geos_doc_build(version, mocked_geos_version, sphinx_doc_build)
     """The requires_geos decorator always adapts the docstring."""
     wrapped = requires_geos(version)(func)
 
-    assert wrapped.__doc__ == expected_docstring.format(version=version, indent=" " * 4)
+    assert wrapped.__doc__ == expected_docstring(version=version, indent=" " * 4)
 
 
 @pytest.mark.parametrize("version", ["3.9.0", "3.10.0"])
@@ -120,7 +130,7 @@ def test_requires_geos_method(version, mocked_geos_version, sphinx_doc_build):
     """The requires_geos decorator adjusts methods docstrings correctly"""
     wrapped = requires_geos(version)(SomeClass.func)
 
-    assert wrapped.__doc__ == expected_docstring.format(version=version, indent=" " * 8)
+    assert wrapped.__doc__ == expected_docstring(version=version, indent=" " * 8)
 
 
 @multithreading_enabled

